packages/circuit-ui/components/Dialog/Dialog.tsx

```diff
@@ -0,0 +1,275 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use client';
+
+import {
+  forwardRef,
+  type HTMLAttributes,
+  type ReactNode,
+  type RefObject,
+  useCallback,
+  useEffect,
+  useRef,
+} from 'react';
+
+import { CloseButton } from '../CloseButton/index.js';
+import dialogPolyfill from '../../vendor/dialog-polyfill/index.js';
+import { applyMultipleRefs } from '../../util/refs.js';
+import { clsx } from '../../styles/clsx.js';
+import {
+  AccessibilityError,
+  isSufficientlyLabelled,
+} from '../../util/errors.js';
+import type { ClickEvent } from '../../types/events.js';
+import { isEscape } from '../../util/key-codes.js';
+
+import classes from './Dialog.module.css';
+import { createUseModalDialog } from './createUseModalDialog.js';
+import {
+  getFirstFocusableElement,
+  hasNativeDialogSupport,
+} from './DialogService.js';
+
+export interface DialogProps
+  extends Omit<HTMLAttributes<HTMLDialogElement>, 'children'> {
+  /**
+   * Whether the dialog is open or not.
+   */
+  open: boolean;
+  /**
+   * Callback when the dialog is closed.
+   */
+  onClose?: () => void;
+  /**
+   * a ReactNode or a function that returns the content of the dialog.
+   */
+  children?:
+    | ReactNode
+    | (({ onClose }: { onClose?: DialogProps['onClose'] }) => ReactNode);
+  /**
+   * Text label for the close button for screen readers.
+   * Important for accessibility.
+   */
+  closeButtonLabel: string;
+  /**
+   * Use the `immersive` variant to focus the user's attention on the dialog content.
+   * default: 'contextual'
+   * */
+  variant?: 'contextual' | 'immersive';
+  /**
+   * Prevent users from closing the modal by clicking/tapping the overlay or
+   * pressing the escape key. Default `false`.
+   */
+  preventClose?: boolean;
+  /**
+   * Enables focusing a particular element in the dialog content and override default behavior
+   */
+  initialFocusRef?: RefObject<HTMLElement>;
+}
+
+export const animationDuration = 300;
+
+export const Dialog = forwardRef<HTMLDialogElement, DialogProps>(
+  (
+    {
+      open,
+      onClose,
+      closeButtonLabel,
+      variant = 'contextual',
+      children,
+      className,
+      preventClose,
+      initialFocusRef,
+      ...props
+    },
+    ref,
+  ) => {
+    const dialogRef = useRef<HTMLDialogElement>(null);
+    const lastFocusedElementRef = useRef<HTMLElement | null>(null);
+    // TODO add translated default close label
+    if (process.env.NODE_ENV !== 'production') {
+      if (!isSufficientlyLabelled(closeButtonLabel)) {
+        throw new AccessibilityError(
+          'Dialog',
+          'The `closeButtonLabel` prop is missing or invalid.',
+        );
+      }
+    }
+    const hasNativeDialog = hasNativeDialogSupport();
+
+    useEffect(() => {
+      const dialogElement = dialogRef.current;
+      if (open && dialogElement) {
+        setTimeout(() => {
```

Comment:
This effect hook should return a cleanup function to clear the timeout in case the component gets unmounted while it's still animating. Otherwise, this would cause a memory leak.